boards/BMS/Src/App/App_PreChargeRelay.c

```diff
@@ -32,3 +32,48 @@ void App_PrechargeRelay_Open(const struct PrechargeRelay *const precharge_relay)
 {
     precharge_relay->open_relay();
 }
+
+void App_PrechargeRelay_IncFaultCounter(struct PrechargeRelay *precharge_relay)
+{
+    precharge_relay->precharge_fault_count++;
+}
+
+uint8_t App_PrechargeRelay_GetFaultCounterVal(const struct PrechargeRelay *const precharge_relay)
+{
+    return precharge_relay->precharge_fault_count;
+}
+
+void App_PrechargeRelay_ResetFaultCounterVal(struct PrechargeRelay *const precharge_relay)
+{
+    precharge_relay->precharge_fault_count = 0U;
+}
+
+bool App_PrechargeRelay_CheckFaults(
+    struct PrechargeRelay *   precharge_relay,
+    struct BmsCanTxInterface *can_tx,
+    bool                      is_charger_connected,
+    bool                      is_ts_rising_slowly,
+    bool                      is_ts_rising_quickly,
+    bool *                    precharge_limit_exceeded)
+{
+    const bool has_precharge_fault =
+        (is_charger_connected) ? is_ts_rising_slowly : (is_ts_rising_slowly | is_ts_rising_quickly);
+
+    if (has_precharge_fault)
+    {
+        App_PrechargeRelay_IncFaultCounter(precharge_relay);
+    }
+
+    if (App_PrechargeRelay_GetFaultCounterVal(precharge_relay) >= 3)
+    {
+        *precharge_limit_exceeded = true;
+    }
+    else
+    {
+        *precharge_limit_exceeded = false;
+    }
```

Comment:
`*precharge_limit_exceeded = App_PrechargeRelay_GetFaultCounterVal(precharge_relay) >= 3`

Reply:
Also nit: You can #define the 3 to `MAX_PRECHARGE_ATTEMPTS` or something like that

Reply:
Oh good call I like that